File: micro-controller-firmware/ros_node/src/odometry_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <cmath>
#include <chrono>

class CarOdometryNode : public rclcpp::Node
{
public:
    CarOdometryNode() : Node("car_odometry_node")
    {
        // Declare parameters
        this->declare_parameter("wheel_radius", 0.03);  // 3cm radius
        this->declare_parameter("wheelbase", 0.185);    // 18.5cm wheelbase
        this->declare_parameter("track_width", 0.15);  // 15cm track width
        this->declare_parameter("publish_rate", 50.0);   // 50Hz
        this->declare_parameter("frame_id", "odom");
        this->declare_parameter("child_frame_id", "base_link");
        
        // Get parameters
        wheel_radius_ = this->get_parameter("wheel_radius").as_double();
        wheelbase_ = this->get_parameter("wheelbase").as_double();
        track_width_ = this->get_parameter("track_width").as_double();
        publish_rate_ = this->get_parameter("publish_rate").as_double();
        frame_id_ = this->get_parameter("frame_id").as_string();
        child_frame_id_ = this->get_parameter("child_frame_id").as_string();
        
        // Initialize odometry state
        x_ = 0.0;
        y_ = 0.0;
        theta_ = 0.0;
        last_time_ = this->now();
        
        // Create publishers
        odom_publisher_ = this->create_publisher<nav_msgs::msg::Odometry>("odom", 10);
        twist_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("odom_twist", 10);
        
        // Create subscribers
        imu_subscription_ = this->create_subscription<sensor_msgs::msg::Imu>(
            "imu/data", 10,
            std::bind(&CarOdometryNode::imu_callback, this, std::placeholders::_1));
            
        motor_subscription_ = this->create_subscription<std_msgs::msg::Float32MultiArray>(
            "motor_data", 10,
            std::bind(&CarOdometryNode::motor_callback, this, std::placeholders::_1));
        
        // Create timer for odometry publishing
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(static_cast<int>(1000.0 / publish_rate_)),
            std::bind(&CarOdometryNode::publish_odometry, this));
        
        // Initialize transform broadcaster
        tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(this);
        
        // Initialize odometry message
        odom_msg_.header.frame_id = frame_id_;
        odom_msg_.child_frame_id = child_frame_id_;
        
        // Set covariance matrices
        set_covariance_matrices();
        
        RCLCPP_INFO(this->get_logger(), "Car Odometry Node started");
        RCLCPP_INFO(this->get_logger(), "Wheel radius: %.3f m", wheel_radius_);
        RCLCPP_INFO(this->get_logger(), "Wheelbase: %.3f m", wheelbase_);
        RCLCPP_INFO(this->get_logger(), "Track width: %.3f m", track_width_);
    }

private:
    // Parameters
    double wheel_radius_;
    double wheelbase_;
    double track_width_;
    double publish_rate_;
    std::string frame_id_;
    std::string child_frame_id_;
    
    // Odometry state
    double x_, y_, theta_;
    rclcpp::Time last_time_;
    
    // Latest sensor data
    sensor_msgs::msg::Imu latest_imu_;
    std_msgs::msg::Float32MultiArray latest_motor_;
    bool imu_received_ = false;
    bool motor_received_ = false;
    
    // ROS2 objects
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_publisher_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr twist_publisher_;
    rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr imu_subscription_;
    rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr motor_subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
    
    // Odometry message
    nav_msgs::msg::Odometry odom_msg_;
    
    void imu_callback(const sensor_msgs::msg::Imu::SharedPtr msg)
    {
        latest_imu_ = *msg;
        imu_received_ = true;
    }
    
    void motor_callback(const std_msgs::msg::Float32MultiArray::SharedPtr msg)
    {
        latest_motor_ = *msg;
        motor_received_ = true;
    }
    
    void set_covariance_matrices()
    {
        // Position covariance (6x6 matrix)
        // [xx, xy, xz, yx, yy, yz, zx, zy, zz]
        odom_msg_.pose.covariance[0] = 0.1;   // xx
        odom_msg_.pose.covariance[7] = 0.1;   // yy
        odom_msg_.pose.covariance[14] = 0.1;   // zz
        odom_msg_.pose.covariance[21] = 0.1;   // roll
        odom_msg_.pose.covariance[28] = 0.1;   // pitch
        odom_msg_.pose.covariance[35] = 0.1;   // yaw
        
        // Velocity covariance (6x6 matrix)
        odom_msg_.twist.covariance[0] = 0.1;   // vx
        odom_msg_.twist.covariance[7] = 0.1;   // vy
        odom_msg_.twist.covariance[14] = 0.1;  // vz
        odom_msg_.twist.covariance[21] = 0.1;  // wx
        odom_msg_.twist.covariance[28] = 0.1;  // wy
        odom_msg_.twist.covariance[35] = 0.1;  // wz
    }
    
    void publish_odometry()
    {
        if (!imu_received_ || !motor_received_) {
            return; // Don't publish if we don't have sensor data
        }
        
        rclcpp::Time current_time = this->now();
        double dt = (current_time - last_time_).seconds();
        
        if (dt <= 0.0) {
            return; // Avoid division by zero
        }
        
        // Calculate velocities from motor data
        double right_rpm = latest_motor_.data[0];
        double left_rpm = latest_motor_.data[1];
        
        // Convert RPM to linear velocities
        double right_velocity = (right_rpm / 60.0) * 2.0 * M_PI * wheel_radius_;
        double left_velocity = (left_rpm / 60.0) * 2.0 * M_PI * wheel_radius_;
        
        // Calculate linear and angular velocities
        double linear_velocity = (right_velocity + left_velocity) / 2.0;
        double angular_velocity = (right_velocity - left_velocity) / track_width_;
        
        // Alternative: Use IMU for angular velocity (more accurate)
        if (imu_received_) {
            angular_velocity = latest_imu_.angular_velocity.z;
        }
        
        // Update position using integration
        x_ += linear_velocity * cos(theta_) * dt;
        y_ += linear_velocity * sin(theta_) * dt;
        theta_ += angular_velocity * dt;
        
        // Normalize angle
        while (theta_ > M_PI) theta_ -= 2.0 * M_PI;
        while (theta_ < -M_PI) theta_ += 2.0 * M_PI;
        
        // Fill odometry message
        odom_msg_.header.stamp = current_time;
        
        // Position
        odom_msg_.pose.pose.position.x = x_;
        odom_msg_.pose.pose.position.y = y_;
        odom_msg_.pose.pose.position.z = 0.0;
        
        // Orientation (convert from yaw to quaternion)
        tf2::Quaternion q;
        q.setRPY(0, 0, theta_);
        odom_msg_.pose.pose.orientation = tf2::toMsg(q);
        
        // Linear velocity
        odom_msg_.twist.twist.linear.x = linear_velocity;
        odom_msg_.twist.twist.linear.y = 0.0;
        odom_msg_.twist.twist.linear.z = 0.0;
        
        // Angular velocity
        odom_msg_.twist.twist.angular.x = 0.0;
        odom_msg_.twist.twist.angular.y = 0.0;
        odom_msg_.twist.twist.angular.z = angular_velocity;
        
        // Publish odometry
        odom_publisher_->publish(odom_msg_);
        
        // Publish twist for debugging
        geometry_msgs::msg::Twist twist_msg;
        twist_msg.linear.x = linear_velocity;
        twist_msg.angular.z = angular_velocity;
        twist_publisher_->publish(twist_msg);
        
        // Publish transform
        publish_transform(current_time);
        
        last_time_ = current_time;
    }
    
    void publish_transform(const rclcpp::Time& time)
    {
        geometry_msgs::msg::TransformStamped transform;
        
        transform.header.stamp = time;
        transform.header.frame_id = frame_id_;
        transform.child_frame_id = child_frame_id_;
        
        transform.transform.translation.x = x_;
        transform.transform.translation.y = y_;
        transform.transform.translation.z = 0.0;
        
        tf2::Quaternion q;
        q.setRPY(0, 0, theta_);
        transform.transform.rotation = tf2::toMsg(q);
        
        tf_broadcaster_->sendTransform(transform);
    }
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<CarOdometryNode>());
    rclcpp::shutdown();
    return 0;
}

File: micro-controller-firmware/ros_node/src/odometry_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from rcl_interfaces.msg import SetParametersResult
from car_state_msg.msg import CarState
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Twist, TransformStamped, Pose
from car_config_msg.msg import CarConfig
from tf2_ros import TransformBroadcaster
import tf2_ros
import tf2_geometry_msgs
import math
import numpy as np
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy


class CarOdometryNode(Node):
    def __init__(self):
        super().__init__("car_odometry_node")

        # Declare parameters - Physical robot parameters
        self.declare_parameter("wheel_radius", 0.0325)
        self.declare_parameter("wheelbase", 0.185)
        self.declare_parameter("track_width", 0.15)
        self.declare_parameter("publish_rate", 50.0)
        self.declare_parameter("frame_id", "odom")
        self.declare_parameter("child_frame_id", "base_link")

        # Velocity filtering parameters
        self.declare_parameter("velocity_threshold", 0.001)  # 1 mm/s
        self.declare_parameter("angular_threshold", 0.001)  # ~0.06 degrees/s

        # Control parameters
        self.declare_parameter(
            "encoder_offset", 187.5
        )  # Steering encoder offset in degrees (0-360)
        self.declare_parameter(
            "max_steering_angle", 30.0
        )  # Maximum steering angle in degrees
        self.declare_parameter("max_rpm", 300.0)  # Maximum motor RPM

        # Debug parameters
        self.declare_parameter("debug_log_frequency", 50)  # Log every Nth message

        # Get parameters
        self.wheel_radius = self.get_parameter("wheel_radius").value
        self.wheelbase = self.get_parameter("wheelbase").value
        self.track_width = self.get_parameter("track_width").value
        self.publish_rate = self.get_parameter("publish_rate").value
        self.frame_id = self.get_parameter("frame_id").value
        self.child_frame_id = self.get_parameter("child_frame_id").value

        # Velocity filtering parameters
        self.velocity_threshold = self.get_parameter("velocity_threshold").value
        self.angular_threshold = self.get_parameter("angular_threshold").value

        # Control parameters
        self.encoder_offset = self.get_parameter("encoder_offset").value
        self.max_steering_angle = self.get_parameter("max_steering_angle").value
        self.max_rpm = self.get_parameter("max_rpm").value

        # Debug parameters
        self.debug_log_frequency = self.get_parameter("debug_log_frequency").value

        # Initialize odometry state
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0
        self.last_time = self.get_clock().now()

        # Latest sensor data
        self.latest_car_state = None
        self.car_state_received = False

        # Create QoS profile for reliable communication
        qos_profile = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            depth=10,
        )

        # Create publishers
        self.odom_publisher = self.create_publisher(Odometry, "odom", 10)
        self.twist_publisher = self.create_publisher(Twist, "odom_twist", 10)
        self.car_config_publisher = self.create_publisher(CarConfig, "car_config", 10)

        # Create subscribers
        self.car_state_subscription = self.create_subscription(
            CarState, "car_state", self.car_state_callback, qos_profile
        )

        # Create timer for odometry publishing (fallback if no data received)
        timer_period = 1.0 / self.publish_rate
        self.timer = self.create_timer(timer_period, self.publish_odometry)

        # Create timer for car config publishing at 1 Hz
        self.config_timer = self.create_timer(1.0, self.publish_car_config)

        # Add parameter change callback for dynamic parameter updates
        self.add_on_set_parameters_callback(self.parameter_change_callback)

        # Initialize transform broadcaster
        self.tf_broadcaster = TransformBroadcaster(self)

        # Initialize odometry message
        self.odom_msg = Odometry()
        self.odom_msg.header.frame_id = self.frame_id
        self.odom_msg.child_frame_id = self.child_frame_id

        # Set covariance matrices
        self.set_covariance_matrices()

        self.get_logger().info("Car Odometry Node started")
        self.get_logger().info(f"Wheel radius: {self.wheel_radius:.3f} m")
        self.get_logger().info(f"Wheelbase: {self.wheelbase:.3f} m")
        self.get_logger().info(f"Track width: {self.track_width:.3f} m")
        self.get_logger().info(f"Encoder offset: {self.encoder_offset:.1f} deg")
        self.get_logger().info(f"Max steering angle: {self.max_steering_angle:.1f} deg")
        self.get_logger().info(f"Max RPM: {self.max_rpm:.1f}")

    def parameter_change_callback(self, params):
        """Handle parameter changes at runtime"""
        for param in params:
            param_name = param.name
            param_value = param.value

            try:
                # Update cached parameter values
                if param_name == "wheel_radius":
                    self.wheel_radius = float(param_value)
                    self.get_logger().info(
                        f"Updated wheel_radius to {self.wheel_radius:.3f} m"
                    )
                elif param_name == "wheelbase":
                    self.wheelbase = float(param_value)
                    self.get_logger().info(
                        f"Updated wheelbase to {self.wheelbase:.3f} m"
                    )
                elif param_name == "track_width":
                    self.track_width = float(param_value)
                    self.get_logger().info(
                        f"Updated track_width to {self.track_width:.3f} m"
                    )
                elif param_name == "encoder_offset":
                    self.encoder_offset = float(param_value)
                    self.get_logger().info(
                        f"Updated encoder_offset to {self.encoder_offset:.1f} deg"
                    )
                elif param_name == "max_steering_angle":
                    self.max_steering_angle = float(param_value)
                    self.get_logger().info(
                        f"Updated max_steering_angle to {self.max_steering_angle:.1f} deg"
                    )
                elif param_name == "max_rpm":
                    self.max_rpm = float(param_value)
                    self.get_logger().info(f"Updated max_rpm to {self.max_rpm:.1f}")
                elif param_name == "velocity_threshold":
                    self.velocity_threshold = float(param_value)
                    self.get_logger().info(
                        f"Updated velocity_threshold to {self.velocity_threshold:.6f} m/s"
                    )
                elif param_name == "angular_threshold":
                    self.angular_threshold = float(param_value)
                    self.get_logger().info(
                        f"Updated angular_threshold to {self.angular_threshold:.6f} rad/s"
                    )
                elif param_name == "publish_rate":
                    old_rate = self.publish_rate
                    new_rate = float(param_value)
                    if new_rate <= 0:
                        return SetParametersResult(
                            successful=False,
                            reason="publish_rate must be greater than 0",
                        )
                    self.publish_rate = new_rate
                    # Update timer period if publish_rate changes
                    timer_period = 1.0 / self.publish_rate
                    self.timer.cancel()
                    self.timer = self.create_timer(timer_period, self.publish_odometry)
                    self.get_logger().info(
                        f"Updated publish_rate from {old_rate:.1f} to {self.publish_rate:.1f} Hz"
                    )
                elif param_name == "frame_id":
                    self.frame_id = str(param_value)
                    self.odom_msg.header.frame_id = self.frame_id
                    self.get_logger().info(f"Updated frame_id to {self.frame_id}")
                elif param_name == "child_frame_id":
                    self.child_frame_id = str(param_value)
                    self.odom_msg.child_frame_id = self.child_frame_id
                    self.get_logger().info(
                        f"Updated child_frame_id to {self.child_frame_id}"
                    )
                elif param_name == "debug_log_frequency":
                    self.debug_log_frequency = int(param_value)
                    self.get_logger().info(
                        f"Updated debug_log_frequency to {self.debug_log_frequency}"
                    )
                else:
                    # Unknown parameter, reject it
                    self.get_logger().warn(f"Unknown parameter: {param_name}")
                    return SetParametersResult(
                        successful=False, reason=f"Unknown parameter: {param_name}"
                    )
            except (ValueError, TypeError) as e:
                self.get_logger().error(
                    f"Invalid value for parameter {param_name}: {e}"
                )
                return SetParametersResult(
                    successful=False, reason=f"Invalid value: {e}"
                )

        # All parameters processed successfully
        return SetParametersResult(successful=True)

    def car_state_callback(self, msg):
        self.latest_car_state = msg
        self.car_state_received = True
        # Publish odometry immediately when new data arrives
        self.publish_odometry()

    def set_covariance_matrices(self):
        # Position covariance (6x6 matrix)
        # [xx, xy, xz, yx, yy, yz, zx, zy, zz]
        self.odom_msg.pose.covariance[0] = 0.1  # xx
        self.odom_msg.pose.covariance[7] = 0.1  # yy
        self.odom_msg.pose.covariance[14] = 0.1  # zz
        self.odom_msg.pose.covariance[21] = 0.1  # roll
        self.odom_msg.pose.covariance[28] = 0.1  # pitch
        self.odom_msg.pose.covariance[35] = 0.1  # yaw

        # Velocity covariance (6x6 matrix)
        self.odom_msg.twist.covariance[0] = 0.1  # vx
        self.odom_msg.twist.covariance[7] = 0.1  # vy
        self.odom_msg.twist.covariance[14] = 0.1  # vz
        self.odom_msg.twist.covariance[21] = 0.1  # wx
        self.odom_msg.twist.covariance[28] = 0.1  # wy
        self.odom_msg.twist.covariance[35] = 0.1  # wz

    def publish_odometry(self):
        if not self.car_state_received:
            return  # Don't publish if we don't have sensor data

        current_time = self.get_clock().now()
        dt = (current_time - self.last_time).nanoseconds / 1e9

        if dt <= 0.0 or dt > 1.0:  # Avoid division by zero and large time jumps
            self.last_time = current_time
            return

        # Calculate velocities from motor data
        right_rpm = self.latest_car_state.right_motor_rpm
        left_rpm = self.latest_car_state.left_motor_rpm

        # Convert RPM to linear velocities (m/s)
        right_velocity = (right_rpm / 60.0) * 2.0 * math.pi * self.wheel_radius
        left_velocity = (left_rpm / 60.0) * 2.0 * math.pi * self.wheel_radius

        # Calculate linear and angular velocities
        linear_velocity = (right_velocity + left_velocity) / 2.0

        # Use IMU for angular velocity (more accurate than wheel difference)
        angular_velocity = self.latest_car_state.gyro_z

        # Apply velocity thresholds to prevent drift (from parameters)
        if abs(linear_velocity) < self.velocity_threshold:
            linear_velocity = 0.0
        if abs(angular_velocity) < self.angular_threshold:
            angular_velocity = 0.0

        # Update position using integration
        self.x += linear_velocity * math.cos(self.theta) * dt
        self.y += linear_velocity * math.sin(self.theta) * dt
        self.theta += angular_velocity * dt

        # Normalize angle
        while self.theta > math.pi:
            self.theta -= 2.0 * math.pi
        while self.theta < -math.pi:
            self.theta += 2.0 * math.pi

        # Fill odometry message
        self.odom_msg.header.stamp = current_time.to_msg()

        # Position
        self.odom_msg.pose.pose.position.x = self.x
        self.odom_msg.pose.pose.position.y = self.y
        self.odom_msg.pose.pose.position.z = 0.0

        # Orientation (convert from yaw to quaternion)
        self.odom_msg.pose.pose.orientation.x = 0.0
        self.odom_msg.pose.pose.orientation.y = 0.0
        self.odom_msg.pose.pose.orientation.z = math.sin(self.theta / 2.0)
        self.odom_msg.pose.pose.orientation.w = math.cos(self.theta / 2.0)

        # Linear velocity
        self.odom_msg.twist.twist.linear.x = linear_velocity
        self.odom_msg.twist.twist.linear.y = 0.0
        self.odom_msg.twist.twist.linear.z = 0.0

        # Angular velocity
        self.odom_msg.twist.twist.angular.x = 0.0
        self.odom_msg.twist.twist.angular.y = 0.0
        self.odom_msg.twist.twist.angular.z = angular_velocity

        # Publish odometry
        self.odom_publisher.publish(self.odom_msg)

        # Publish twist for debugging
        twist_msg = Twist()
        twist_msg.linear.x = linear_velocity
        twist_msg.angular.z = angular_velocity
        self.twist_publisher.publish(twist_msg)

        # Publish transform
        self.publish_transform(current_time)

        # Debug logging (frequency from parameter)
        if hasattr(self, "debug_counter"):
            self.debug_counter += 1
        else:
            self.debug_counter = 0

        if self.debug_counter % self.debug_log_frequency == 0:
            self.get_logger().info(
                f"Odometry: x={self.x:.3f}, y={self.y:.3f}, theta={self.theta:.3f}, "
                f"vx={linear_velocity:.3f}, wz={angular_velocity:.3f}, dt={dt:.3f}, "
                f"RPM: R={right_rpm:.1f}, L={left_rpm:.1f}"
            )

        self.last_time = current_time

    def publish_transform(self, time):
        transform = TransformStamped()

        transform.header.stamp = time.to_msg()
        transform.header.frame_id = self.frame_id
        transform.child_frame_id = self.child_frame_id

        transform.transform.translation.x = self.x
        transform.transform.translation.y = self.y
        transform.transform.translation.z = 0.0

        transform.transform.rotation.x = 0.0
        transform.transform.rotation.y = 0.0
        transform.transform.rotation.z = math.sin(self.theta / 2.0)
        transform.transform.rotation.w = math.cos(self.theta / 2.0)

        self.tf_broadcaster.sendTransform(transform)

    def publish_car_config(self):
        """Publish car configuration at 1 Hz based on parameters"""
        config_msg = CarConfig()

        # Physical parameters
        config_msg.wheelbase = self.wheelbase
        config_msg.track_width = self.track_width
        config_msg.wheel_radius = self.wheel_radius

        # Encoder configuration and control limits from parameters
        config_msg.encoder_offset = self.encoder_offset
        config_msg.max_steering_angle = self.max_steering_angle
        config_msg.max_rpm = self.max_rpm

        self.car_config_publisher.publish(config_msg)


def main(args=None):
    rclpy.init(args=args)
    node = CarOdometryNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
